### PR TITLE
User point static

### PR DESCRIPTION
**Pull request by ravifrancesco**
Made the UserPoint class a singleton class. Now it's accessible through the static method `UserPoint.getInstance(Context mContext)`. 

Adapted tests accordingly

### CODE REVIEW

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/ComputePOIPointsTest.java [hunk 1]
```diff
@@ -24,11 +24,14 @@
...
     private static UserPoint userPoint;
+    private static Context mContext;
 
     @BeforeClass
     public static void setUserPoint(){
         //Set userPoint to the mont blanc
-        userPoint = new UserPoint(45.802537, 6.850328, 4809);
+        mContext = ApplicationProvider.getApplicationContext();
+        userPoint = UserPoint.getInstance(mContext); // TODO check if working properly
+        userPoint.setLocation(46.519251915333676, 6.558563221333525, 220, 0);
```
Comment (magangloff) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/ComputePOIPointsTest.java:34]: You could rechange location to the mont-blanc and change the altitude or maybe change the comment above

**Review by magangloff — CHANGES_REQUESTED** [review]
Small comment for the test, else all good.

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java [hunk 2]
```diff
@@ -65,9 +66,12 @@
...
         Context context = ApplicationProvider.getApplicationContext();
         Assert.assertNotNull(context);
-        UserPoint userPoint = new UserPoint(MOCK_LOCATION_LAT_LAUSANNE,
-                                            MOCK_LOCATION_LON_LAUSANNE,
-                                            MOCK_LOCATION_ALT_LAUSANNE);
+        UserPoint userPoint = UserPoint.getInstance(context);
+        userPoint.setLocation(MOCK_LOCATION_LAT_LAUSANNE,
+                MOCK_LOCATION_LON_LAUSANNE,
+                MOCK_LOCATION_LON_LAUSANNE,
+                0);
```
Comment (bgabriel998) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java:73]: Error in function call, third parameter should be altitude
```suggestion
        userPoint.setLocation(MOCK_LOCATION_LAT_LAUSANNE,
                MOCK_LOCATION_LON_LAUSANNE,
                MOCK_LOCATION_ALT_LAUSANNE,
                0);
```

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java [hunk 3]
```diff
@@ -222,7 +226,8 @@ public void testGeonamesObjCreationExceptionShortRange() throws IllegalArgumentE
         thrown.expectMessage("BoundingBoxRangeKm can't be null or negative (also not under 100m)");
         Context context = ApplicationProvider.getApplicationContext();
         Assert.assertNotNull(context);
-        UserPoint userPoint = new UserPoint(46.519251915333676,6.558563221333525,220);
+        UserPoint userPoint = UserPoint.getInstance(context);
+        userPoint.setLocation(46.519251915333676, 6.558563221333525, 220, 0);
```
Comment (bgabriel998) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java:230]: Use constants instead
```suggestion
        userPoint.setLocation(MOCK_LOCATION_LAT_LAUSANNE, MOCK_LOCATION_LON_LAUSANNE, MOCK_LOCATION_ALT_LAUSANNE, 0);
```

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java [hunk 4]
```diff
@@ -251,7 +256,8 @@ public void testGeonamesObjCreationExceptionQueeryMaxAsZero() throws IllegalArgu
 
         Context context = ApplicationProvider.getApplicationContext();
         Assert.assertNotNull(context);
-        UserPoint userPoint = new UserPoint(46.519251915333676,6.558563221333525,220);
+        UserPoint userPoint = UserPoint.getInstance(context);
+        userPoint.setLocation(46.519251915333676, 6.558563221333525, 220, 0);
```
Comment (bgabriel998) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java:260]: Use constants instead
```suggestion
        userPoint.setLocation(MOCK_LOCATION_LAT_LAUSANNE, MOCK_LOCATION_LON_LAUSANNE, MOCK_LOCATION_ALT_LAUSANNE, 0);
```

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java [hunk 5]
```diff
@@ -279,7 +285,8 @@ public void testGeonamesObjCreationExceptionTimeoutAsZero() throws IllegalArgume
         thrown.expectMessage("QueryTimeout parameter can't be less than 1 sec");
         Context context = ApplicationProvider.getApplicationContext();
         Assert.assertNotNull(context);
-        UserPoint userPoint = new UserPoint(46.519251915333676,6.558563221333525,220);
+        UserPoint userPoint = UserPoint.getInstance(context);
+        userPoint.setLocation(46.519251915333676, 6.558563221333525, 220, 0);
```
Comment (bgabriel998) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/GeonamesHandlerTest.java:289]: Use constants instead
```suggestion
        userPoint.setLocation(MOCK_LOCATION_LAT_LAUSANNE, MOCK_LOCATION_LON_LAUSANNE, MOCK_LOCATION_ALT_LAUSANNE, 0);
```

File: app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/ComputePOIPointsTest.java [hunk 6]
```diff
@@ -24,11 +24,14 @@
...
     public GrantPermissionRule grantCameraPermissionRule = GrantPermissionRule.grant(Manifest.permission.CAMERA);
     private static UserPoint userPoint;
+    private static Context mContext;
 
     @BeforeClass
     public static void setUserPoint(){
         //Set userPoint to the mont blanc
-        userPoint = new UserPoint(45.802537, 6.850328, 4809);
+        mContext = ApplicationProvider.getApplicationContext();
+        userPoint = UserPoint.getInstance(mContext); // TODO check if working properly
```
Comment (bgabriel998) [inline app/src/androidTest/java/com/github/bgabriel998/softwaredevproject/points/ComputePOIPointsTest.java:33]: Remove TODO

```suggestion
        userPoint = UserPoint.getInstance(mContext); 
```

**Review by bgabriel998 — APPROVED** [review]
Change magic numbers for constants. 
Otherwise the code is fine. The singleton is a good decision

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java [hunk 7]
```diff
@@ -29,24 +35,33 @@
...
+     * initialized the parameter will be ignored.
+     *
+     * @param mContext  context of the application.
+     * @return          single instance of the user point.
+     */
+    public static UserPoint getInstance(Context mContext) {
+        if (single_instance == null) {
+            return new UserPoint(mContext);
+        }
+        return single_instance;
```
Comment (magangloff) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java:59]: Can be changed for better readability:
```suggestion
        return single_instance == null ? new UserPoint(mContext) : single_instance;
```

**Review by bgabriel998 — APPROVED** [review]

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java [hunk 8]
```diff
@@ -18,7 +22,9 @@
  *
  * This class should be used as a observer that observes a GPSTracker.
  */
-public class UserPoint extends Point {
+public final class UserPoint extends Point {
+
+    private static final UserPoint single_instance = null; // singleton instance
```
Comment (magangloff) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java:27]: Should not be made final to assign a variable to single_instance
```suggestion
    private static UserPoint single_instance = null; // singleton instance
```

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java [hunk 9]
```diff
@@ -29,24 +35,33 @@
...
+     * Constructor for the UserPoint. Private because it is a singleton.
      *
      * @param mContext  the current context of the app
      */
-    public UserPoint(Context mContext) {
+    private UserPoint(Context mContext) {
         super(GPSTracker.DEFAULT_LAT, GPSTracker.DEFAULT_LON, GPSTracker.DEFAULT_ALT);
         gpsTracker = new GPSTracker(mContext, this);
         customLocation = false;
     }
```
Comment (magangloff) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/points/UserPoint.java:46]: ```single_instance``` should be assigned here, else ```single_instance``` is always ```null```
```suggestion
        single_instance = this;
    }
```

**Review by magangloff — CHANGES_REQUESTED** [review]
```single_instance``` in UserPoint was never updated and thus always ```null```. This should be updated.

**Review by magangloff — APPROVED** [review]

**Pull request merged**